README.md
# banner_widget
------
A new Flutter  BannerWidget project 


------

## How to Use
```python
# add this line to your dependencies
count_down_timer: ^0.0.1
```


```python
 BannerWidget(data,
                duration: 2000,
                height: 200.0,
                selectedColor: Colors.red,
                unSelectedColor: Colors.white,
                onBannerItemClick: (int position, BannerItem item) {
                  Fluttertoast.showToast(
                      msg: 'index=$position banner is clicked！',
                      toastLength: Toast.LENGTH_LONG);
                },
                build: null,
                descriptionBackgroundColor: const Color(0x33000000),
                textInfoDirect: InfoDirect.HORIZONTAL,
                circleRadius: 5.0,
                indicatorStyle: IndicatorStyle.ELLIPTICAL,
                ellipticalWidth: 16.0,
                ellipticalHeight: 8.0,
                cornerRadius: 10.0),
```
### API

| property        | description    | defaultValue
| :--------   | :-----  |:--------
| duration     | int    | 2000
| height        | double | 200.0
| selectedColor        |Color | Colors.red |
| unSelectedColor        | Color | Colors.white |
| onBannerItemClick        | OnBannerItemClick(int position, BannerItem bannerItem)   | null
| build        | CustomBuild(int position, BannerItem item) | null
|descriptionBackgroundColor   | Color | const Color(0x33000000)
|textInfoDirect   | InfoDirect |  InfoDirect.HORIZONTAL
|circleRadius   | double | 5.0
|indicatorStyle   | IndicatorStyle | IndicatorStyle.ELLIPTICAL
|ellipticalWidth   | double | 16.0
|ellipticalHeight   | double | 8.0
|cornerRadius   | double | 10.0




###  [GitHub](https://github.com/manburenshenglu/banner_widget)

CHANGELOG.md
* TODO: Describe initial release.
## [0.0.1] - TODO: Add release date.
init commit

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:banner_widget/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/widget/banner_widget.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:transparent_image/transparent_image.dart';

const MAX_COUNT = 0x7fffffff;
enum InfoDirect { HORIZONTAL, VERTICAL }
enum IndicatorStyle {
  ///circle indicator
  CIRCLE,

  ///elliptical indicator
  ELLIPTICAL
}

///banner click callback
typedef void OnBannerItemClick(int position, BannerItem bannerItem);

typedef Widget CustomBuild(int position, BannerItem item);

///author xjl
///description
///date 2019/7/16 17:29
///modify
class BannerWidget extends StatefulWidget {
  final int duration;
  final double height;
  final List<BannerItem> data;
  final Color selectedColor;
  final Color unSelectedColor;
  final double circleRadius;

  ///elliptical indicator width
  final double ellipticalWidth;

  ///elliptical indicator height
  final double ellipticalHeight;

  ///elliptical indicator cornerRadius
  final double cornerRadius;
  final IndicatorStyle indicatorStyle;

  ///text background color
  final Color descriptionBackgroundColor;
  final OnBannerItemClick onBannerItemClick;
  final CustomBuild build;

  final InfoDirect textInfoDirect;

  BannerWidget(this.data,
      {this.duration: 2000,
      this.height: 200.0,
      this.selectedColor: Colors.red,
      this.unSelectedColor: Colors.white,
      this.onBannerItemClick,
      this.build,
      this.descriptionBackgroundColor: const Color(0x33000000),
      this.textInfoDirect: InfoDirect.HORIZONTAL,
      this.circleRadius: 5.0,
      this.indicatorStyle: IndicatorStyle.CIRCLE,
      this.ellipticalWidth: 16.0,
      this.ellipticalHeight: 8.0,
      this.cornerRadius: 10.0})
      : assert(data != null);

  @override
  _BannerWidgetState createState() => _BannerWidgetState();
}

class _BannerWidgetState extends State<BannerWidget> {
  Timer _timer;

  PageController _pageController;

  ///current selected banner index
  int selectedIndex = 0;

  @override
  void initState() {
    // TODO: implement initState
    double current = widget.data.length > 0
        ? (MAX_COUNT / 2) - ((MAX_COUNT / 2) % widget.data.length)
        : 0.0;
    _pageController = PageController(initialPage: current.toInt());
    start();
    super.initState();
  }

  @override
  void didChangeDependencies() {
    // TODO: implement didChangeDependencies
    super.didChangeDependencies();
  }

  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return Container(
      height: widget.height,
      color: Colors.black12,
      child: Stack(
        children: <Widget>[
          getPageView(),
          Align(
            alignment: Alignment.bottomCenter,
            child: IntrinsicHeight(
              child: Container(
                padding: EdgeInsets.all(5.0),
                color: widget.descriptionBackgroundColor,
                child: getBannerTextInfoWidget(),
              ),
            ),
          )
        ],
      ),
    );
  }

  void onPageChanged(index) {
    selectedIndex = index % widget.data.length;
    setState(() {});
  }

  ///return banners
  Widget getPageView() {
    return PageView.builder(
        itemCount: widget.data.length > 0 ? MAX_COUNT : 0,
        controller: _pageController,
        onPageChanged: onPageChanged,
        itemBuilder: (context, index) {
          return InkWell(
              onTap: () {
                if (widget.onBannerItemClick != null) {
                  widget.onBannerItemClick(
                      selectedIndex, widget.data[selectedIndex]);
                }
              },
              child: widget.build == null
                  ? FadeInImage.memoryNetwork(
                      placeholder: kTransparentImage,
                      image: widget.data[index % widget.data.length].imgPath,
                      fit: BoxFit.cover)
                  : widget.build(
                      index, widget.data[index % widget.data.length]));
        });
  }

  Widget getBannerTextInfoWidget() {
    if (widget.textInfoDirect == InfoDirect.VERTICAL) {
      return Column(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            getCurrentTextInfoWidget(),
            IntrinsicWidth(child: Row(children: indicator()))
          ]);
    } else {
      return Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Expanded(flex: 1, child: getCurrentTextInfoWidget()),
          Expanded(
              flex: 0,
              child: Row(
                  mainAxisSize: MainAxisSize.min, children: indicator()))
        ],
      );
    }
  }

  List<Widget> indicator() {
    List<Widget> indicator = [];
    for (int i = 0; i < widget.data.length; i++) {
      indicator.add(
        Container(
            margin: EdgeInsets.all(3.0),
            width: widget.indicatorStyle == IndicatorStyle.CIRCLE
                ? widget.circleRadius
                : widget.ellipticalWidth,
            height: widget.indicatorStyle == IndicatorStyle.CIRCLE
                ? widget.circleRadius
                : widget.ellipticalHeight,
            decoration: BoxDecoration(
                borderRadius: widget.indicatorStyle == IndicatorStyle.CIRCLE
                    ? null
                    : BorderRadius.all(Radius.elliptical(
                        widget.cornerRadius, widget.cornerRadius)),
                shape: widget.indicatorStyle == IndicatorStyle.CIRCLE
                    ? BoxShape.circle
                    : BoxShape.rectangle,
                color: i == selectedIndex
                    ? widget.selectedColor
                    : widget.unSelectedColor)),
      );
    }
    return indicator;
  }

  Widget getCurrentTextInfoWidget() {
    return widget.data.length > 0 && selectedIndex < widget.data.length
        ? widget.data[selectedIndex].textInfoWidget
        : Text('$selectedIndex');
  }

  @override
  void didUpdateWidget(BannerWidget oldWidget) {
    // TODO: implement didUpdateWidget
    super.didUpdateWidget(oldWidget);
  }

  void stop() {
    _timer?.cancel();
    _timer = null;
  }

  void start() {
    stop();
    _timer = Timer.periodic(Duration(milliseconds: widget.duration), (timer) {
      if (widget.data.length > 0 &&
          _pageController != null &&
          _pageController.page != null) {
        _pageController.animateToPage(_pageController.page.toInt() + 1,
            duration: Duration(milliseconds: 300), curve: Curves.linear);
      }
    });
  }

  @override
  void dispose() {
    // TODO: implement dispose
    stop();
    super.dispose();
  }
}

class BannerItem {
  String imgPath;
  Widget textInfoWidget;

  static BannerItem defaultBannerItem(String image, String text,
      [maxLines = 1,
      style = const TextStyle(
          color: Colors.white,
          fontSize: 12.0,
          decoration: TextDecoration.none)]) {
    BannerItem item = new BannerItem();
    item.imgPath = image;
    Text textWidget = new Text(
      text,
      softWrap: true,
      maxLines: maxLines,
      overflow: TextOverflow.ellipsis,
      style: style,
    );

    item.textInfoWidget = textWidget;

    return item;
  }
}


lib/main.dart
import 'package:banner_widget/widget/banner_widget.dart';
import 'package:flutter/material.dart';
import 'package:fluttertoast/fluttertoast.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  List<BannerItem> data = [];

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    BannerItem item = BannerItem.defaultBannerItem(
        'http://n.sinaimg.cn/sports/transform/0/w500h300/20190814/14d8-icapxpi4070373.jpg',
        'LeBron James');
    data.add(item);
    item = BannerItem.defaultBannerItem(
        'http://img.mp.itc.cn/upload/20170106/a87363f5c7e548ec8022b5cadfc1c216_th.jpg',
        'LeBron&Wade');
    data.add(item);
    item = BannerItem.defaultBannerItem(
        'http://02.imgmini.eastday.com/mobile/20180328/20180328212318_1ef989a9c001edab4c94578141c10f18_2.jpeg',
        'LeBron VS Wade');
    data.add(item);
  }

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.start,
          children: <Widget>[
            BannerWidget(data,
                duration: 2000,
                height: 200.0,
                selectedColor: Colors.red,
                unSelectedColor: Colors.white,
                onBannerItemClick: (int position, BannerItem item) {
                  Fluttertoast.showToast(
                      msg: 'index=$position banner is clicked！',
                      toastLength: Toast.LENGTH_LONG);
                },
                build: null,
                descriptionBackgroundColor: const Color(0x33000000),
                textInfoDirect: InfoDirect.HORIZONTAL,
                circleRadius: 5.0,
                indicatorStyle: IndicatorStyle.ELLIPTICAL,
                ellipticalWidth: 16.0,
                ellipticalHeight: 8.0,
                cornerRadius: 10.0),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
